### ISSUE TITLE
Feature Request: Ability to specify arguments before subcommands

### ISSUE THREAD

**Issue by LordMike** [natemcmaster/CommandLineUtils#265]
**Is your feature request related to a problem? Please describe.**
I'd like to create a commandline that allows me to do this:

```
program.exe -d value alpha // -d is currently ignored
program.exe -d value beta // -d is currently ignored
program.exe alpha -d value
program.exe beta -d value
```

That is.. the ability to specify some argument before a subcommand, to specify some common argument. My subcommands are currently of an inherited form, where the base class has arguments such as "Database path", and I'd like that to take affect no matter where it was placed.

The odd thing is, even if I specify the "-d" before my subcommand, no exception is thrown - but the argument is discarded. I noticed because my database hadn't been modified .. in the place I expected.

**Describe the solution you'd like**
Either a parsing exception should be thrown, or the value should be used. In no case should the value be silently ignored.

**Describe alternatives you've considered**
My workaround is to remember to put the subcommand first, always.

I vaguely recall other tools have the opposite issue - they will ignore common arguments that go after some "action". The dotnet CLI seems to also exhibit this, although they'll warn you about it.

**Comment by natemcmaster** [comment]
Is this the behavior you're trying to achieve?

```csharp
using System;
using McMaster.Extensions.CommandLineUtils;

[Subcommand(typeof(Alpha), typeof(Beta))]
class Program
{
    static void Main(string[] args)
        => CommandLineApplication.Execute<Program>(args);

    [Option(Inherited = true)]
    public string D { get; }
}

class Alpha
{
    Program Parent { get; }
    public void OnExecute()
        => Console.WriteLine($"alpha (d = {Parent.D})");
}

class Beta
{
    Program Parent { get; }
    public void OnExecute()
        => Console.WriteLine($"beta (d = {Parent.D})");
}

```

**Comment by LordMike** [comment]
Am I simply missing `inherited` ? :O

**Comment by LordMike** [comment]
Ooh.. You're also injecting the parent. I've inherited from the parent (Program in your case).

**Comment by LordMike** [comment]
While attempting to reproduce/make-it-work/.. I've encountered some errors in  `2.3.4`.

This example gives a Stack Overflow Exception when run without arguments, and when given `-d a alpha`
```csharp
[Subcommand(typeof(Alpha), typeof(Beta))]
class Program
{
    static void Main(string[] args)
        => CommandLineApplication.Execute<Program>(args);

    [Option(Inherited = true)]
    public string D { get; }
}

class Alpha : Program
{
    Program Parent { get; }

    public void OnExecute()
        => Console.WriteLine($"alpha (d = {Parent.D})");
}

class Beta : Program
{
    Program Parent { get; }

    public void OnExecute()
        => Console.WriteLine($"beta (d = {Parent.D})");
}

```

**Comment by natemcmaster** [comment]
Uh oh, stackoverflows are no fun. I'll take a look soon, hopefully this weekend

**Comment by JakeStaTeresa** [comment]
@LordMike,  You're getting a stack overflow exception because Alpha and Beta both inherit from Program. The attribute resolution reads the SubCommand attribute from the parent first and then it goes to each type specified and does the same. But because the SubCommand attribute is inherited, parent and child effectively has the same attribute. And the cycle goes...

**Comment by LordMike** [comment]
Right. A helpful exception should be thrown (track all seen types, if one repeats, throw an exception indicating the recursiveness).

**Comment by JakeStaTeresa** [comment]
This is already fixed in master at #239

**Comment by vpkopylov** [comment]
> 
> 
> This is already fixed in master at #239 and originally reported s #236
> 
> As the cyclic dependency is a manifestation of improper use. And fixing the cause makes the issue go away. I think its ok to wait until v2.4.0 to get the proper error message.

I confirm that SubcommandCycleException is thrown when running this code on master

**Comment by natemcmaster** [comment]
Thanks for confirming! I'm closing this one as resolved.
